content/posts/tclap.md

```diff
@@ -0,0 +1,74 @@
+---
+title: "Using TCLAP to parse dumpsys commands"
+date: 2023-09-06T17:49:38+02:00
+draft: false
+author: "The Android Team"
+---
+
+## Introduction
+The `Templatized C++ Command Line Parser` is a c++ header library that is used to parse command line arguments. The developers of this library maintain a manual (https://tclap.sourceforge.net/manual.html) that is easy to use, so be sure to check it out.
+
+## Usage
+In order to use this library to parse your dumpsys commands, you can just copy the following code and adapt the todos to fit your needs
+
+```
+binder_status_t <your_class>::dump(int fd, const char **argv, uint32_t argc) {
+
+    // because dumpsys is printing to a custom file descriptor and tclap uses cout and cerr
+    // to do the printing, we have to redirect everything to a local buffer that we will print
+    // at the end of this function. Because of this limitation, this function must have only
+    // 1 return point (otherwise, we need to copy-paste code around)
+    ::std::stringstream buffer;
+    ::std::streambuf *old_cout = ::std::cout.rdbuf(buffer.rdbuf());
+    ::std::streambuf *old_cerr = ::std::cerr.rdbuf(buffer.rdbuf());
+
+    try {
+          ::TCLAP::CmdLine cmd("<short_description_of_what_this_command_line_is_for>", ' ', "1.0", true);
+        // configure commands
+
+        // we need to handle all exceptions ourselves, because, if TCLAP finds an exception, it will
+        // also invoke the "exit" function which will terminate the process that contains this instance
+        // of TCLAP
+        cmd.setExceptionHandling(false);
+
+        // because TCLAP was meant as a command line tool that expects the program name as argv[0]
+        // and we receive the parameters without the program name, it is necessary that we insert
+        // the program name as argv[0] so we do not have to change the TCLAP library
+        ::std::vector<std::string> argsAsVect;
+        argsAsVect.push_back("<name_of_your_service_that_appears_after_dumpsys>");
+        for (uint32_t i = 0; i < argc; ++i) {
+            ::std::string next_argv(argv[i]);
+            argsAsVect.push_back(next_argv);
+        }
+
+        // todo: add the possible commands that can be invoked
+        // (see https://tclap.sourceforge.net/manual.html)
+
+        cmd.parse(argsAsVect);
+
+        // todo: process the commands
+
+    } catch (::TCLAP::ArgException &e) {
+        ::std::cerr << "error: " << e.error() << " for arg " << e.argId() << ::std::endl;
+    } catch (const ::TCLAP::CmdLineParseException &e) {
+        std::cerr << e.what() << std::endl;
+    } catch (const ::TCLAP::ExitException &e) {
+    }
+
+    dprintf(fd, "%s", buffer.str().c_str());
+    ::std::cerr.rdbuf(old_cerr); // reset to standard error again
+    ::std::cout.rdbuf(old_cout); // reset to standard output again
+    return EX_NONE;
+}
+```
+
+## Constraints
```

Comment:
Consider "Android specifics" instead of "Constraints", just because "Constrains" sounds so negative ;-)

Reply:
done